pkg/apis/admission/types.go

```diff
@@ -0,0 +1,160 @@
+}
+
+// Initializer describes the name and the failure policy of an initializer.
+type Initializer struct {
+	// Name is the identifier of the initializer. It will be added to the
+	// object that needs to be initialized.
+	Name string
+
+	// FailurePolicy defines what happens if there is an initializer controller
+	// fails to takes action. Allowed values are Ignore, or Fail. If "Ignore" is
+	// set, initializer is removed from the initializers list of an object if
+	// the timeout is reached; If "Fail" is set, apiserver returns timeout error
+	// if the timeout is reached.
+	FailurePolicy *FailurePolicyType
+}
+
+type FailurePolicyType string
+
+const (
+	// Ignore means the initilizer is removed from the initializers list of an
+	// object if the initializer is timed out.
+	Ignore FailurePolicyType = "Ignore"
+	// For 1.7, only "Ignore" is allowed. "Fail" will be allow when the
+	// extensible admission feature is more mature.
+	Fail FailurePolicyType = "Fail"
+)
+
+// ExternalAdmissionHook describes an external admission webhook and the
+// resources and operations it applies to.
+type ExternalAdmissionHook struct {
+	// The name of the external admission webhook.
+	Name string
+
+	// ClientConfig defines how to communicate with the hook.
+	ClientConfig AdmissionHookClientConfig
+
+	// Rules describes what operations on what resources/subresources the webhook cares about.
+	// The webhook cares about an operation if it matches _any_ Rule.
+	Rules []Rule
+
+	// FailurePolicy defines how unrecognized errors from the admission endpoint are handled -
+	// allowed values are Ignore or Fail. Defaults to Ignore.
+	FailurePolicy FailurePolicyType
+}
+
+// Rule describes the Verbs and Resources an admission hook cares about. Each
+// Rule is a tuple of Verbs and Resources.It is recommended to make sure all
+// the tuple expansions are valid.
+type Rule struct {
+	// Verbs is the verbs the admission hook cares about - CREATE, UPDATE, or *
+	// for all verbs.
+	// Defaults to '*'.
+	// If '*' is present, the length of the slice must be one.
+	Verbs []OperationType
+
+	// APIGroups is the API groups the resources belong to. '*' is all groups.
+	// Defaults to '*'
+	// If '*' is present, the length of the slice must be one.
+	APIGroups []string
+
+	// APIVersions is the API versions the resources belong to. '*' is all versions.
+	// Defaults to '*'.
+	// If '*' is present, the length of the slice must be one.
+	APIVersions []string
+
+	// Resources is a list of resources this rule applies to.
+	//
+	// For example:
+	// 'pods' means pods.
+	// 'pods/log' means the log subresource of pods.
+	// '*' means all resources, but not subresources.
+	// 'pods/*' means all subresources of pods.
+	// '*/scale' means all scale subresources.
+	// '*/*' means all resources and their subresources.
+	//
+	// Defaults to '*'.
+	// If '*' or '*/*' is present, the length of the slice must be one.
+	Resources []string
+}
+
+type OperationType string
+
+const (
```

Comment:
Since this is no longer a mutating AC and is basically there for validation and approval, only supporting `CREATE` and `UPDATE` seems like an oversight.  For example, I might want to disallow the deletion of network policy objects because doing this might break multi-tenancy security.